includes/blocks/course-theme/class-complete-lesson.php

```diff
@@ -0,0 +1,95 @@
+<?php
+/**
+ * File containing the Complete_Lesson class.
+ *
+ * @package sensei
+ * @since
+ */
+
+namespace Sensei\Blocks\Course_Theme;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+use \Sensei_Blocks;
+
+/**
+ * Class Complete_Lesson is responsible for rendering the 'Mark complete' block.
+ */
+class Complete_Lesson {
+
+	/**
+	 * Complete_Lesson constructor.
+	 */
+	public function __construct() {
+		Sensei_Blocks::register_sensei_block(
+			'sensei-lms/course-theme-complete-lesson',
+			[
+				'render_callback' => [ $this, 'render' ],
+			]
+		);
+	}
+
+	/**
+	 * Renders the block.
+	 *
+	 * @access private
+	 *
+	 * @return string The block HTML.
+	 */
+	public function render() : string {
+		$post = get_post();
+
+		// Return empty if we can't get the post for some reason.
+		if ( empty( $post ) ) {
+			return '';
+		}
+
+		$lesson_id = $post->ID;
+		$course_id = \Sensei()->lesson->get_course_id( $lesson_id );
+
+		if (
+			// Return empty if it is not a lesson post.
+			'lesson' !== $post->post_type ||
+
+			// Return empty if user is not enrolled.
+			! \Sensei_Course::is_user_enrolled( $course_id ) ||
+
+			// Return empty if the lesson requires a quiz pass.
+			\Sensei()->lesson->lesson_has_quiz_with_questions_and_pass_required( $lesson_id ) ||
+
+			// Return empty if user already completed the lesson.
+			\Sensei_Utils::user_completed_lesson( $lesson_id )
+		) {
+			return '';
+		}
+
+		// The button is disabled if the lesson requires a pre-requisite.
+		$disabled = '';
+		if ( ! \Sensei_Lesson::is_prerequisite_complete( $lesson_id, get_current_user_id() ) ) {
+			$disabled = 'disabled';
+		}
+
+		// The button is a secondary CTA if there is a quiz but not required to take/pass it.
+		$secondary = '';
+		if ( \Sensei_Lesson::lesson_quiz_has_questions( $lesson_id ) ) {
+			$secondary = 'sensei-course-theme-complete-lesson__secondary';
+		}
+
+		// Render "Mark Complete" button.
+		$nonce     = wp_nonce_field( 'woothemes_sensei_complete_lesson_noonce', 'woothemes_sensei_complete_lesson_noonce', false, false );
+		$permalink = esc_url( get_permalink() );
+		$text      = esc_html( __( 'Complete lesson', 'sensei-lms' ) );
+
+		return ( "
+			<form class='sensei-course-theme-complete-lesson-form' method='POST' action='{$permalink}'>
```

Comment:
Here I have that same observation about the single quotes (since we use double quotes for the HTML in all the other parts, it would be nice to keep the same standard), but we can change that when moving it to the actions block! =)

Reply:
Updated! 👍 

Btw, what would be the reason we're following this convention? Why HTML attribute values should be double quoted instead of single quotes?

Reply:
FWIW here's the [WP Coding Standards section](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/#quotes) on quotes for HTML attributes. According to the WP Standard, quotes are necessary, but single vs double doesn't matter.

Reply:
> what would be the reason we're following this convention?

Honestly, I don't have the answer to why we're following double quotes as rule in the plugin. I think just because it was the preference of the team that started writing the plugin. :)

Reply:
> FWIW here's the WP Coding Standards section on quotes for HTML attributes. According to the WP Standard, quotes are necessary, but single vs double doesn't matter.

> Honestly, I don't have the answer to why we're following double quotes as rule in the plugin. I think just because it was the preference of the team that started writing the plugin. :)

Well in that case, moving forward I think it should be OK to wrap our html with double quotes and use single quotes for attributes. It's much better to write and read that way.

#### Double quote HTML attributes

```php
$html = '
  <form class="sensei-course-theme-complete-lesson-form" method="POST" action="' . $permalink . '">
    ' . $nonce . '
    <input type="hidden" name="quiz_action" value="lesson-complete" />
    <button type="submit" class="sensei-course-theme-complete-lesson ' . $secondary . '" ' . $disabled . '>
      ' . $text . '
    </button>
  </form>
';
```

#### Single quote HTML attributes

```php
$html = "
  <form class='sensei-course-theme-complete-lesson-form' method='POST' action='{$permalink}'>
    {$nonce}
    <input type='hidden' name='quiz_action' value='lesson-complete' />
    <button type='submit' class='sensei-course-theme-complete-lesson{$secondary}' {$disabled}>
      {$text}
    </button>
  </form>
";
```

Declarative FTW!

Reply:
That would be okay for me too! I think it's just a question of setting a convention for our codebase. =) Maybe it would be a topic to open a discussion on a P2 and define?

Also, one thing I thought a while ago was to create templates for any HTML we write in PHP (maybe with the templates we have currently - that the user can override, or a template lib just to separate the markup). So we would avoid writing HTML inside PHP.